crates/ruff_linter/src/rules/refurb/rules/hardcoded_string_charset.rs

```diff
@@ -0,0 +1,168 @@
+use crate::checkers::ast::Checker;
+use crate::importer::ImportRequest;
+use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{CmpOp, Expr, ExprStringLiteral, StringLike};
+use ruff_text_size::{Ranged, TextRange};
+
+/// ## What it does
+/// Checks for uses of hardcoded charsets, which are defined in Python string module.
+///
+/// ## Why is this bad?
+/// Usage of named charsets from the standard library is more readable and less error-prone.
+///
+/// ## Example
+/// ```python
+/// x = "0123456789"
+/// y in "abcdefghijklmnopqrstuvwxyz"
+/// ```
+///
+/// Use instead
+/// ```python
+/// import string
+///
+/// x = string.digits
+/// y in string.ascii_lowercase
+/// ```
+///
+/// ## References
+/// - [Python documentation: String constants](https://docs.python.org/3/library/string.html#string-constants)
+#[violation]
+pub struct HardcodedStringCharset {
+    name: &'static str,
+}
+
+impl AlwaysFixableViolation for HardcodedStringCharset {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Use of hardcoded string charset")
+    }
+
+    fn fix_title(&self) -> String {
+        let HardcodedStringCharset { name } = self;
+        format!("Replace hardcoded charset with `string.{name}`")
+    }
+}
+
+struct Charset {
+    name: &'static str,
+    bytes: &'static [u8],
+    bitset: u128,
+}
```

Comment:
This is clever! But also somewhat complicated. Have you done a performance comparison with https://docs.rs/aho-corasick/latest/aho_corasick/ 

CC: @BurntSushi maybe you have a sense for whether the bitset approach or corasick is better here.

Reply:
> Have you done a performance comparison with https://docs.rs/aho-corasick/latest/aho_corasick/?

No.

There are two parts here:
- `exact`: there are 9 `&[u8]` comparisons, which are usually fail by `.len()` comparison. Do not believe that aho-corasick could help a lot (if we do not lift the scope higher than `fn string_like`).
- `as_set`: 
  - in the current implementation works only in `... [not] in <str literal>` case (looks not so often to care)
  - if we would like to save the semantic `set(...) == set(string.<charset>)`, AFAIK, aho-corasick does not help us. We still need some `set`-ish solution (e.g. like this inlined bitset). The one idea we can add is "early return if the byte is not `printable`", but looks like premature optimization for me.

Reply:
I don't have a good sense here. 9 memcmp's could be very quick. This is the sort of thing where, if you want to go as fast as possible, you probably need a micro-benchmark to try and optimize it. But I think the approach here looks very reasonable to me.